File: ES_Toolkit_R/R/GridRequestor.R
#' Get climate data grids for specified parameter(s) and NPS unit(s)
#'
#' Takes one park code and a list of one or more climate parameters and requests climate data for the specified duration (daily, monthly, or yearly). Returns a grid or grids (by parameter) in ASCII format or an in-memory raster stack.
#' @param unitCode (optional) One unit code as a string. If not specified, defaults to CONUS geographic extent.
#' @param sDate sdate (required) Format as a string (yyyy-mm, yyyymm, yyyy). The beginning of the desired date range.
#' @param eDate edate (required) Format as a string (yyyy-mm, yyyymm, yyyy). The end of the desired date range.
#' @param distance (optional) Distance (in kilometers) to buffer park bounding box
#' @param climateParameters (optional) A list of one or more daily, monthly, or yearly climate parameters (e.g. pcpn, mint, maxt, avgt or mly_pcpn, mly_mint, mly_maxt, mly_avgt or yly_pcpn, yly_mint, yly_maxt, yly_avgt).  If not specified, defaults to all parameters for the specified duration. See the ACIS Web Services page: http://www.rcc-acis.org/docs_webservices.html
#' @param duration (optional)  "dly" | "mly" | "yly" (i.e, daily, monthly, or yearly). If not specified, defaults to dly.
#' @param filePath filePath (optional) Folder path for output ASCII grid(s). If specified, grid(s) are saved to the folder. Otherwise, grid(s) are saved to an in-memory raster stack (by date).
#' @return ASCII-formatted grid file for each parameter for each date or an in-memory raster stack with a layer for each parameter for each date
#' @examples \dontrun{
#' Two daily grids for GRSM for one date: print output to console
#' getGrids(unitCode = list("GRSM"), sdate = "20160615", edate = "20160616", climateParameters = list("mint", "maxt"), duration = "dly")
#' 
#' Two monthly grids for PRWI for one date: returns one grid for each parameter for each date - 4 grids total
#' getGrids(unitCode = list("PRWI"), sdate = "201606", edate = "201607", duration = "mly", climateParameters = list("mly_mint", "mly_maxt"), filePath="d:\\temp\\trash")
#'
#' Two monthly grids for GRSM for one date: print output to console
#' getGrids(unitCode = list("GRSM"), sdate = "20160615", edate = "20160616", climateParameters = list("mly_mint", "mly_maxt"), duration = "mly")
#' 
#' The same request for two monthly grids for GRSM for one date: output to in-memory raster stack, t
#' t <- getGrids(unitCode = list("GRSM"), sdate = "201606", edate = "201607", duration = "mly", climateParameters = list("mly_mint", "mly_maxt"))
#' 
#' Get all monthly grids for one year for one parameter
#' getGrids(unitCode = list("GRKO"), sdate = "190001", edate = "190012", duration = "mly", climateParameters = list("mly_mint"))
#' 
#' Get yearly grids for three years for all parameters
#' getGrids(unitCode = list("SHEN"), sdate = "2013", edate = "2016", duration = "yly")
#' }
#' @export
#' 
getGrids <-
  function (unitCode = NULL,
            sdate = NULL,
            edate = NULL,
            distance = NULL,
            climateParameters = NULL,
            duration = "dly",
            filePath = NULL) {
    # URLs and request parameters:
    # ACIS data services
    baseURL <- "http://data.rcc-acis.org/"
    webServiceSource <- "GridData"
    config <- add_headers(Accept = "'Accept':'application/json'")
    
    # Default to CONUS extent
    if (is.null(unitCode)) {
      bbox <- "-130, 20,-50, 60"
    } else {
      # NPS Park bounding box
      if (is.null(distance)) {
        bboxExpand  = 0.0
      } else {
        bboxExpand = distance * 0.011 # convert km to decimal degrees
      }
      bbox <- getBBox(unitCode, bboxExpand)
    }
    
    # Initialize body of request
    bList <- list(bbox = bbox,
                  sdate = sdate,
                  edate = edate)
    
    # Hard-coded request elements
    gridElements <-
      list(
        #interval = "dly",
        duration = duration,
        gridSource = "PRISM",
        dataPrecision = 1,
        output = "json",
        meta = "ll"
      )
    
    # Image parameters from lookup file - used for output formatting and  request documentation
    lookups <- 
      fromJSON(system.file("ACISLookups.json", package = "EnviromentalSettingToolkit"), flatten = TRUE) # assumes placement in package inst subfolder
    luElements  <- lookups$gridSources[gridElements$gridSource]
    
    
    # Configure image output
    fileNameRoot <- unlist(gridElements$gridSource)
    gridElements <- c(gridElements, grid = luElements[[1]]$code)
    if (is.null(filePath) == TRUE) {
      rasterStack <- stack()
    }
    
    bList <- c(bList, output = unlist(gridElements$output))
    bList <- c(bList, grid = unlist(gridElements$grid))
    bList <- c(bList, meta = unlist(gridElements$meta))
    
    # If climateParameters is NULL, default to these parameters
    if (is.null(climateParameters) == TRUE) {
      if (duration == "dly") {
        #climateParameters <- list('pcpn', 'mint', 'maxt', 'avgt')
        climateParameters <- lookups$elementGrid$code
      }
      else {
        # Monthly (mly) or Yearly (yly)
        climateParameters <- lookups$elementAggregateGrid$code[lookups$elementAggregateGrid$duration == gridElements$duration]
      }
    }
    
    # Format incoming arguments
    dataURL <-  gsub(" ", "", paste(baseURL, webServiceSource))
    #climateParameters <- c(paramElements)
    climateElems <- paste(climateParameters, collapse = ",")
    paramCount <- length(climateParameters)
    
    # Iterate parameter list to create elems element:
    eList <- vector('list', paramCount)
    for (i in 1:paramCount) {
      e <-
        list(
          name = unlist(c(climateParameters[i])),
          #interval = gridElements$interval,
          duration = gridElements$duration,
          prec = gridElements$dataPrecision
        )
      #print(e)
      eList[[i]] <- e
    }
    
    # Climate parameters as JSON
    elems <- toJSON(eList, auto_unbox = TRUE)
    body  <- stripEscapes(c(bList, elems = elems))
    print(dataURL)
    print(body)
    
    # Format POST request for use in httr
    # Initialize response object
    dfResponse <- NULL
    # This returns the full response - need to use content() and parse
    dataResponseInit <-
      POST(
        dataURL,
        accept_json(),
        add_headers("Content-Type" = "application/json"),
        body = body,
        verbose()
      )
    
    # Format climate data object
    rList <- content(dataResponseInit)
    dataResponseError <- rList$error
    if (is.null(dataResponseError)) {
      resp <-
        fromJSON(content(dataResponseInit, "text"), simplifyVector = FALSE)
      #respJSON  <- toJSON(content(dataResponseInit, "text"), auto_unbox = TRUE)
      for (i in 1:length(resp$data)) {
        # Convert each output to ASCII format - grid data is resp$data[[i]][[2]]) as nested list
        print(resp$data[[i]][[1]]) #image date
        # Get the minimum grid center value
        minLon <- min(as.numeric(unlist(resp$meta$lon)))
        minLat <- min(as.numeric(unlist(resp$meta$lat)))
        maxLon <- max(as.numeric(unlist(resp$meta$lon)))
        maxLat <- max(as.numeric(unlist(resp$meta$lat)))
        #print(length(resp$data[[i]][[2]])) #count of image rows
        # Get grid data as a matrix; output to file or console
        for (j in 1:length(climateParameters)) {
          fileName <- NULL
          sepDuration <- paste(paste("_", gridElements$duration, sep=""), "_", sep="")
          fileName <-
            paste(paste(
              paste(fileNameRoot, climateParameters[j], sep = "_"),
              resp$data[[i]][[1]],
              sep = sepDuration#"_dly_"
            ), ".asc", sep = "")
          #fileName <- cat(paste(paste(fileNameRoot, climateParameters[j], sep = "_"), resp$data[[i]][[1]], sep = "_dly_"), ".asc")
          #print(fileName)
          gridMatrix <- do.call(rbind, resp$data[[i]][[j + 1]])
          griddf <- NULL
          for (k in 1:ncol(gridMatrix)) {
            # for proper image orientation, flip matrix columns with rev()
            griddf <-
              cbind(griddf, as.numeric(rev(gridMatrix[, k])))
          }
          if (is.null(filePath) == FALSE) {
            #print(griddf)
            outfile <-
              outputAscii(griddf,
                          paste(filePath, fileName, sep = "\\"),
                          minLon,
                          minLat,
                          luElements[[1]])
            if (outfile == "Success") {
              print(
                cat(
                  "SUCCESS: Created raster(s) for",
                  unlist(unitCode),
                  "using climateParameter=",
                  unlist(climateParameters)[j],
                  "in " ,
                  filePath
                )
              )
            }
            else {
              print(
                cat(
                  "ERROR: Unable to create raster(s) for ",
                  unlist(unitCode),
                  "using climateParameter=",
                  unlist(climateParameters)[j]
                )
              )
            }
          }
          else {
            print("INFO: No filePath specified. Output written to console:\n")
            outfile <-
              outputAscii(griddf, "", minLon, minLat, luElements[[1]])
            if (outfile == "Success") {
              #outputRaster <- outputRasterStack()
              r <- raster(griddf, xmn=minLon, xmx=maxLon, ymn=minLat, ymx=maxLat)
              crs(r) <- luElements[[1]]$projectionCRS
              names(r) <- paste(resp$data[[i]][[1]], unlist(climateParameters)[j])
              rasterStack <- stack(rasterStack, r)
            }
            else {
              print(
                cat(
                  "ERROR: Unable to create raster(s) for ",
                  unlist(unitCode),
                  "using climateParameter=",
                  unlist(climateParameters)[j]
                )
              )
            }
          }
        } # param loop end
      } # date loop end
      if (exists("rasterStack") == TRUE) { 
        return (rasterStack)
      }
    }
    else {
      print(paste("ERROR: ", dataResponseError))
    }
  }

File: ES_Toolkit_py/GridRequestor.py
from ACIS import ACIS
from common import formatStringArguments
from GridStack import GridStack

class GridRequestor(ACIS):
    gridSource = 'PRISM' #ONLY PRISM IS SUPPORTED AT PRESENT
    precision = 1 #precision is only set for gridded data at present
    add = None
    reduceCodes = []

    def __init__(self):
        super(GridRequestor,self).__init__()
        self.defaultParameters = self.supportedParameters
        self.duration = None
        self.webServiceSource = 'GridData'

    @property
    def supportedParameters(self):
        '''
        Supported parameters for grids - this will need to be updated as we
        introduce more grid sources
        '''
        #NOTE: These parameters are specific to PRISM
        return ['pcpn', 'mint', 'maxt', 'avgt']

    def _callForGrids(self, **kwargs):
        '''
        Core method to request grids from ACIS
        '''
        #add the metadata and  precision level to the set of arguments
        kwargs = self._formatArguments(kwargs, meta = 'll', precision = self.precision)

        #Based on the grid source, get additional metadata values specific to grids
        gridSourceCode = self.gridSources[self.gridSource]['code']
        missingValue = int(self.gridSources[self.gridSource]['missingValue'])
        cellSize = float(self.gridSources[self.gridSource]['cellSize'])
        projection = self.gridSources[self.gridSource]['projection']

        #call ACIS for grids and extract out the latitude and longitude values
        #from the response
        grids = self._call_ACIS(kwargs, grid = gridSourceCode)
        self._checkResponseForErrors(grids)
        latValues = grids['meta']['lat']
        lonValues = grids['meta']['lon']

        #Instantiate grid stack object
        gs = GridStack(gridSource = self.gridSource, latValues = latValues, lonValues = lonValues, cellSize = cellSize,
                projection = projection, aggregation = self.interval, missingValue = missingValue)

        #iterate through all of the grids and add them to the stack
        for grid in grids['data']:
            for index, variable in enumerate(self.climateParameters):
                gs._addGrid(variable = variable, date = grid[0].encode(), grid = grid[index+1])

        #If file Path and name are provided, export all grids
        if self.filePath:
            gs.export(filePath = self.filePath)
        return gs

    def getGrids(self, sdate, edate, duration, unitCode = None, distance = 0,
        climateParameters = None, filePath = None):
        '''
        Method to fetch daily grids from ACIS.  Currently only PRISM grids are
        supported.

        ARGUMENTS
            sdate               Start date (yyyy-mm-dd or yyyymmdd).
            edate               End date (yyyy-mm-dd or yyyymmdd).
            duration            dly | mly | yly (i.e, daily, monthly, or yearly)
            unitCode (optional) 4-letter unit code. Currently accepts only one.
            distance (optional) Distance in kilometers for buffering a bounding box of park.
                                If no distance is specified then 0 is used as the default buffer.
            climateParameters (optional)    Accepts one or more of the climate parameter codes,
                                            preferably as a list or tuple
            filePath (optional)             If provided, one or more ascii grids are saved to the
                                    working directory. Grid names follow the pattern of
                                    Source_parameter_aggregation_YYYYMMDD (e.g., PRISM_mint_dly_20150101)

        RETURNS
            Dictionary like object (aka GridStack) containing one or more daily grids.
            Grids are indexed first by parameter and then by date
        '''
        self.interval = duration
        self.duration = duration


        if self.duration != 'dly':
            climateParameters = formatStringArguments(climateParameters, self.supportedParameters)
            climateParameters = map(lambda p: self.duration + '_' + p, climateParameters)

        return self._callForGrids(unitCode = unitCode, sdate = sdate, edate = edate
            , distance = distance, filePath = filePath, climateParameters = climateParameters,
            reduceCodes = self.reduceCodes, prec = self.precision )


if __name__ == '__main__':
    gr = GridRequestor()
    filePath = 'C:\\TEMP\\'

    ##YEARLY GRIDS

    #TEST 01
    sdate = '1915'
    edate = '1929'
    climateParameters = ['mint', 'maxt']
    unitCode = 'LIBO'
    distance = 10

    data =  gr.getGrids(sdate = sdate, edate = edate, duration = 'yly'
        ,unitCode = unitCode, distance = distance,
        climateParameters = climateParameters, filePath = filePath )
    print data.climateParameters
    print data.dates

    ##MONTHLY GRIDS

    #TEST 01
    sdate = '1895-01'
    edate = '1896-12'
    climateParameters = ['mint', 'maxt']
    unitCode = 'YELL'
    distance = 0

    data =  gr.getGrids(sdate = sdate, edate = edate, duration = 'mly'
        ,unitCode = unitCode, distance = distance,
        climateParameters = climateParameters, filePath = filePath )
    print data.climateParameters
    print data.dates
    data.export(filePath = filePath)
    print data['mly_mint']['1895-01']
    print data.dates
    print data.climateParameters
    data['mly_mint']['1895-01'].export(filePathAndName = filePath + 'test.asc')
##
    #Test 02
    unitCode = 'OLYM'
    sdate = '20150115'
    edate = '20150615'
    climateParameters = ['maxt']
    distance = 0
    data =  gr.getGrids(sdate = sdate, edate = edate, duration = 'mly'
        ,unitCode = unitCode, distance = distance,
        climateParameters = climateParameters, filePath = filePath )


    ##DAILY GRIDS
##    #TEST 01
##    sdate = '2015-01-01'
##    edate = '2015-01-04'
##    climateParameters = 'mint, maxt'
##    unitCode = 'APPA'
##    distance = 0
##
##    data =  gr.getGrids(sdate = sdate, edate = edate, duration = 'dly',
##        unitCode = unitCode, distance = distance,
##        climateParameters = climateParameters, filePath = filePath )
##    print data.climateParameters
##    print data.dates
##    data.export(filePath = filePath)
##    print data['mint']['2015-01-03']
##    print data.dates
##    print data.climateParameters
##    data['mint']['2015-01-03'].export(filePathAndName = filePath + 'test.asc')
##
##    #Test 02
##    unitCode = 'OLYM'
##    sdate = '20150615'
##    edate = '20150615'
##    climateParameters = ['maxt']
##    distance = 0
##    data =  gr.getGrids(sdate = sdate, edate = edate, duration = 'dly',
##        unitCode = unitCode, distance = distance,
##    climateParameters = climateParameters, filePath = filePath )



